MicroService code re-factoring

vkuznet opened the pull request:
This PR revisits MicroService code and made it working with proposed [MS-Transferor](https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor) architecture. So far it does the following:
- talks to ReqMgr2 and gets `acquired` workflows
- talks to Workqueue and gets GQE
- performs `as is` Unified Transferor logic

vkuznet committed:
Revisit MicroService code and made it working

amaltaro commented on src/python/WMCore/MicroService/Unified/Common.py:
```diff
@@ -259,60 +276,71 @@ def cert():
-    default = 'http://cmsmonitoring.web.cern.ch/cmsmonitoring'
-    return uConfig.get('monitoringUrl', default)
+    if not url:
+        url = 'http://cmsmonitoring.web.cern.ch/cmsmonitoring'
```
Don't waste time refactoring these simple get methods. Eventually we should convert all of them to service arguments and that's it.

amaltaro commented on src/python/WMCore/MicroService/Unified/Common.py:
```diff
@@ -81,6 +85,19 @@ def phedexInfo(datasets):
             blockNodes[item['name']] = nodes
     return blockNodes
 
+def getRequests(status):
```
This is not using what I suggested here: https://github.com/dmwm/WMCore/issues/9142#issuecomment-482565706

vkuznet replied:
Alan, this is the part of **Unified black box** logic which I didn't touch yet. I want you first to concentrate on parts which I made for first release, and once it is everything in place I'll move on fixing stuff like that.

vkuznet replied:
Ohh, sorry, I thought about something else. I'll fix it.

vkuznet commented:
Alan, I fixed this API.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> amaltaro commented on this pull request.
> 
> 
> 
> > @@ -81,6 +85,19 @@ def phedexInfo(datasets):
>              blockNodes[item['name']] = nodes
>      return blockNodes
>  
> +def getRequests(status):
> 
> This is not using what I suggested here: https://github.com/dmwm/WMCore/issues/9142#issuecomment-482565706
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9140#pullrequestreview-229364565

amaltaro commented:
Valentin, I noticed this week that the model we have been discussing was missing one use case. Which is regarding the pileup sample and the need to have it completely available for a workflow (either in one or in scattered among many sites). Which brings me to the conclusion that we cannot have a service only making subscriptions and not monitoring them.

I'm drafting another proposal and will write it down soon, such that we can go through it later today, if you have time.

amaltaro commented:
@vkuznet I'd suggest to separate the developments you're putting into this PR into a couple of PRs, or maybe even 3. This PR has a) code refactoring, b) further developments for Transferor and c) implementation of TransferMonit, making it harder to validate and thus easier to introduce a bug.

If not too late, try to at least keep those 3 above in different commits.

vkuznet commented:
Alan,
since both MSTransferor and MSTransferorMonit are integrated it would be hard to
separate this PR because we want to have fully working system at once.
Instead, I suggest to merge this PR and then separate further
development into different streams as you suggested. Moreover it seems
logical since now we can concentrate on individual implementations.
V.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> @vkuznet I'd suggest to separate the developments you're putting into this PR into a couple of PRs, or maybe even 3. This PR has a) code refactoring, b) further developments for Transferor and c) implementation of TransferMonit, making it harder to validate and thus easier to introduce a bug.
> 
> If not too late, try to at least keep those 3 above in different commits. 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9140#issuecomment-493834921

vkuznet commented:
@amaltaro , Alan please review this PR. I removed WorkQueue pieces from the code as was requested on today's meeting.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+        """
+        try:
+            # get requests from ReqMgr2 data-service for given statue
+            requestSpecs = self.svc.reqmgr.getRequestByStatus([req_status], detail=True)
```
We don't need request details here (detail=False). We only need to know workflow names in this status, all the rest of the information should be available from the application database containinig a map of workflow -- dataset -- transfer id

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+            self.logger.debug('+++ monit found %s requests in %s state' % (len(requests), req_status))
+
+            # nothing to do
+            if not requests:
```
these 3 lines are not really needed, the API should return an empty dict, so the for loop below will be automatically skipped.

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+        # https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
+        statuses = []
+        records = self.store.records()
+        store_tids = [r.get('tid', 0) for r in records]
```
You are scanning this `records` structure twice. Remove this line above and anything related to `store_tids`, just find the tid from the records as done in the for loop below.

vkuznet replied:
I will not do it until we'll settle on couchdb record formats and you'll implement proper APIs. So far it is placeholder code which allow to implement required functionality. Its logic will change once we'll get proper tids and APIs.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+        tids = [r['tid'] for r in self.store.records()]
+        return tids
+
+    def getRequest(self, tid):
```
It should be the other way around, return transfer id for a given request (sort of as implemented above).

vkuznet replied:
ok, I remove this method (i used it for testing purposes).

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+                    if rec.get('tid', 0) == tid:
+                       statuses.append(rec['status'])
+                       break
+            else:
```
For the record, this else should never get executed. The MS is supposed to persist all transfer ids, so if we cannot find it, we have a bug.

vkuznet replied:
once again, it is placeholder code to make whole workflow working. Once proper APIs will be in place it will be re-written. Moreover, most likely I'll remove internal store as well.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+                return rec['request']
+
+    def getTransferStatuses(self, tids):
+        "get transfer ids from backend"
```
update this doc string

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
         # return percentage of completion
         return round(float(sum(sdict.values()))/float(tot), 2) * 100
 
+    def checkTransferRequests(self, request):
```
I'm not sure I understand the difference between these 3 methods: getTransferStatuses, checkSubscription and this one. To me they seem to have the same goal, checking the status of the transfer requests from the cache or querying the data-service to get up-to-date status.

vkuznet replied:
so far it is placeholder. We may need all 3 for different purposes:
- getTransferStatuses will fetch doc from couchdb and return transfer statuses
- checkSubscription will ask upstream server
- checkTransferRequest will use both above if necessary

I don't think we should worry right now about not implemented function, instead you should evaluate place where it is used.

amaltaro replied:
I just want to avoid with code that "we may need" that is actually never needed and just adds up to code maintenance and complexity. In general, I'd say don't implement something unless you're sure it will be needed (this is something I've been fighting in WMCore over these many past years).

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
+                % (prefix, request, rec, req_status))
+            if rec['reqStatus'] != req_status:
+                self.svc.reqmgr.updateRequestStatus(request, req_status)
+                self.store.update(request, {'reqStatus': req_status})
```
I'd rather not save the request status in multiple databases. We know sooner or later it goes out of sync and it just gives us more problems. I'd say we don't store request status in this self.store temp db.

vkuznet replied:
I agree, but since there is no yet couchdb one I used internal store to test whole system. As I wrote in previous comments it is most likely will be gone once we'll have couchdb one.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
 
+		# TODO: when we made subscription we change status of request in ReqMgr2
+
     def info(self, request=None):
```
Shouldn't `request` parameter be mandatory here? If it's not provided, what exactly is it returning?

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
+class Services(object):
+    "Services class provides access to reqmgr2 services: ReqMgr, ReqMgrAux"
+    __slots__ = ('reqmgrAux', 'reqmgr', 'workqueue')
+    def __init__(self, reqmgrUrl, workqueueUrl, logger=None):
```
Remove workqueueUrl.
BTW, I'd completely remove this class and just get an instance of ReqMgr, ReqMgrAux, wherever you need it (apparently it's used only once for now)

vkuznet replied:
my oversight, I agree and will remove it.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
+
 class UnifiedTransferorManager(object):
     """
     Initialize UnifiedTransferorManager class
```
Can you improve this doc string with what this class is meant to do? Does it deserve its own python file?

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
+        if logger:
+            self.logger = logger
+        else:
+            self.logger = logging.getLogger('reqmgr2ms:UnifiedTransferorManager')
```
How about replacing it by
`logging.getLogger(self.__class__.__name__)`
?

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
+            self.logger = logging.getLogger('reqmgr2ms:UnifiedTransferorManager')
+            logging.basicConfig()
+        reqmgrUrl = getattr(config, 'reqmgrUrl', 'https://cmsweb.cern.ch/reqmgr2')
+        workqueueUrl = getattr(config, 'workqueueUrl', 'https://cmsweb.cern.ch/couchdb/workqueue')
```
no need for this

vkuznet replied:
removed

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -168,37 +384,56 @@ def add(self, requests):
+        # update uConfig urls according reqmgr2ms configuration
+        uConfig.set('reqmgrUrl', reqmgrUrl)
+        uConfig.set('reqmgrCacheUrl', getattr(config, 'reqmgrCacheUrl', 'https://cmsweb.cern.ch/couchdb/reqmgr_workload_cache'))
+        uConfig.set('workqueueUrl', getattr(config, 'workqueueUrl', workqueueUrl))
```
no need again

vkuznet replied:
removed

amaltaro reviewed (changes requested):
I made many comments along the code.
In addition to that, I have a few comments and concerns to share:
* given that most of the web-services interaction are already implemented in WMCore, it makes sense to use those Services API when possible; however, we need to look out for performance degradation, since those don't support a pool of requests to be made all in parallel as the pycurl does.
* I'm certainly missing some details, but I'm afraid we're not caching many things that could, thus making many unnecessary duplicate calls (e.g., fetching requests from ReqMgr)
* you have a mix of variable naming convention all over this code. Let's stick to camelCase please.

I have to say this code is also still very confusing to me. Reason is likely because we copied over lots of things from unified. So whatever we can make simpler, we should!

vkuznet commented on setup_dependencies.py:
```diff
@@ -75,8 +75,11 @@
         'modules': ['WMCore.Wrappers.__init__',
                     'WMCore.Wrappers.JsonWrapper.__init__',
                     'WMCore.Wrappers.JsonWrapper.JSONThunker',
+                    'WMCore.WorkQueue.__init__',
```
yes this can be removed now.

vkuznet commented on src/python/WMCore/MicroService/Unified/Common.py:
```diff
@@ -259,60 +254,71 @@ def cert():
     return pair[1]
 
-def stucktransferUrl():
+def stucktransferUrl(url=None):
```
since I already made all of these changes, I'll leave them here.

vkuznet commented on src/python/WMCore/MicroService/Unified/RequestInfo.py:
```diff
@@ -9,19 +9,19 @@
     reqmgrCacheUrl, dbsUrl, eventsLumisInfo, workflowsInfo, \
     dbsInfo, phedexInfo, reqmgrUrl, elapsedTime, getComputingTime, \
-    getNCopies, teraBytes, workqueueView
+    getNCopies, teraBytes, getWorkflow
```
I agree.

vkuznet commented on src/python/WMCore/MicroService/Unified/RequestInfo.py:
```diff
@@ -335,41 +342,89 @@ def getSiteWhiteList(request, siteInfo, reqmgrAuxSvc, reqSpecs=None, pickone=Fal
+            requestsToProcess[name] = requestSpecs[name]
+    return requestsToProcess
+
+def requestsInfo_old(svc, state='acquired', requests=None, logger=None, verbose=False):
```
I would like to keep this code since it contains original port of Unified logic. Once I'll port it to requestsInfo we'll remove this function.

vkuznet commented on src/python/WMCore/MicroService/Unified/SiteInfo.py:
```diff
@@ -81,13 +81,14 @@ def fetch(self):
-            '%s/stuck_m6.json' % stucktransferUrl(),
+# VK: so far this is unavailable
+#            '%s/stuck_1.json' % stucktransferUrl(),
+#            '%s/stuck_2.json' % stucktransferUrl(),
```
ok, will remove

vkuznet commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -72,88 +86,290 @@ def close(self):
         self.verbose = verbose
+        if verbose:
+            self.logger.setLevel(logging.DEBUG)
         if not dbFileName:
```
no, you're wrong, the DB file name is configuration parameter we can and will make it persistent.

amaltaro replied:
Then there is no need to make this parameter optional. But I guess you want to keep it for testing purposes, for the moment(?)

amaltaro commented:
@vkuznet did you manage to run the very basic tests here? I understand it depends on many other things, but is the code at least not crashing as it is?

The concerns I expressed in my previous comment are still valid; and we will need to revisit them in the next PR.

I'm merging a bunch of things by tomorrow (cherrypy changes, new request status, pulling workqueue elements from another status) and possibly this one such that we can see how it goes in testbed.

Can you please squash all non **test** changes into the same commit and leave unittests/pylint changes in a second commit? Thanks

amaltaro commented on test/python/WMCore_t/MicroService_t/Unified_t/Transferor_t.py:
```diff
@@ -16,16 +16,18 @@
         RequestStore, RequestManager
```
And that's why jenkins report 7 tests deleted. Because you're importing `RequestManager` here, which you renamed to `MSManager`. Please fix it.

vkuznet commented:
Alan,
thanks for update. I fixed unix tests flow even though they need separate revision since logic has changed completely. I squashed all commits to move things forward.

Feel free to review/decide tomorrow if you want to merge it.

vkuznet commented:
Alan,
please merge this PR that I can move forward. Honestly I spent more time on comments, then on actual work. I already made significant clean-up of this code, see
https://gist.github.com/vkuznet/ba5fd2015963ed12a21029c9df02e04f
where I remove local store, I removed APIs you complained, I put clear logic based on proposal. But I want to apply changes once this PR is merged. This will allow me to come back to code and re-use it if necessary.

amaltaro commented:
I guess you're just not used to have people reviewing your PRs.
Alright, let's put it in. Next PR will be the production-quality one though :) Thanks!

amaltaro commented on src/python/WMCore/MicroService/Unified/RequestInfo.py:
```diff
@@ -306,7 +309,11 @@ def getSiteWhiteList(request, siteInfo, reqmgrAuxSvc, reqSpecs=None, pickone=Fal
+        # res = svc.reqmgrAux.postCampaignConfig(campaign, {'%s_name' % campaign: {"Key1": "Value1"}})
+        campaignConfig = svc.reqmgrAux.getCampaignConfig(campaign)
+        if isinstance(campaignConfig, list):
+           campaignConfig = campaignConfig[0]
```
one empty space less here btw. Pylint report gives us many good insights!

vkuznet commented:
It is not that, I doubt that we both need to waste time on code which is far
from what we need and bother about pre-mature optimization/quality.

Once we achieve the functional code according to proposal then reviewing PR in
terms of code syntax and other stuff makes sense.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> I guess you're just not used to have people reviewing your PRs.
> Alright, let's put it in. Next PR will be the production-quality one though :) Thanks!
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9140#issuecomment-495354546

amaltaro commented:
I disagree. Not having a final and functional code doesn't mean we can push in garbage code full of very simple mistakes that can be fixed. Most of code convention, unused imports/vars,  can be fixed with a couple of clicks in an IDE. So the cost is very small to have a production-quality code in the repo. Last but not least, we know that half of those cases we never get back to the code and improve it.

vkuznet commented:
Alan,
we may have disagreement on how to approach code/repositories. But I value my
and peoples time. As such, while awaiting a simple merge I spent time to adjust
code to new specs. This version is more close to specs, it addressed many issues
you raised.  But since you're resisted to merge intermediate code, I saved it for
on my own and pushed new one forward. Feel free to review again.
Valentin.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> I disagree. Not having a final and functional code doesn't mean we can push in garbage code full of very simple mistakes that can be fixed. Most of code convention, unused imports/vars,  can be fixed with a couple of clicks in an IDE. So the cost is very small to have a production-quality code in the repo. Last but not least, we know that half of those cases we never get back to the code and improve it.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9140#issuecomment-495363838
